### PR TITLE
[SPARK-21115][Core]If the cores left is less than the coresPerExecutor,the cores left will not be allocated, so it should not to check in every schedule

### PR DESCRIPTION
**Pull request by eatoncys**
## What changes were proposed in this pull request?
If we start an app with the param --total-executor-cores=4 and spark.executor.cores=3, the cores left is always 1, so it will try to allocate executors in the function org.apache.spark.deploy.master.startExecutorsOnWorkers in every schedule.
Another question is, is it will be better to allocate another executor with 1 core for the cores left.

## How was this patch tested?
unit test

### CODE REVIEW

**Comment by jerryshao** [comment]
According to my test, current code of `Master` could handle this situation correctly, did you see any issue here without your fix?

**Comment by eatoncys** [comment]
@jerryshao I have not see any issue here, and I have tested this again using the latest Master code, the problem also exists.

**Comment by jerryshao** [comment]
If you don't see any issue here, what's problem you met?

**Comment by eatoncys** [comment]
@jerryshao The problem is: If we start an app with the param --total-executor-cores=4 and spark.executor.cores=3,  the code "app.coresLeft>0" is always true in "org.apache.spark.deploy.master.startExecutorsOnWorkers" and it will try to allocate executor for this app and it will allocate nothing, it is better to compare the app.coresLeft whih coresPerExecutor, if the coresLeft less than coresPerExecutor, it will return directly.

**Comment by eatoncys** [comment]
@jerryshao I have modified the "app.coresLeft>0" to "app.coresLeft >= coresPerExecutor.getOrElse(1)".
And another question is :  is it will be better to allocate another executor with 1 core for the cores left?

**Comment by jerryshao** [comment]
I see, I understand your changes now.

IMO, because user specifically request for 3 cores per executor (as an example), it is not so good to allocate 1 executor with only 1 core, this may break user's purpose. Also the original behavior of `Master` is to not allocate executor with residual resources, we should keep the convention.

Besides, I think it would be better to add some warning logs in `SparkSubmit` when total cores is not divisible by cores per executor.

**Comment by eatoncys** [comment]
@jerryshao Ok, I will add warning logs in SparkSubmit, thanks.

**Comment by eatoncys** [comment]
@jerryshao I have added warning logs in SparkSubmit , would you like to review it again, thanks.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala [hunk 1]
```diff
@@ -278,6 +278,15 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
     if (pyFiles != null && !isPython) {
       SparkSubmit.printErrorAndExit("--py-files given but primary resource is not a Python script")
     }
+    if (totalExecutorCores != null && executorCores != null) {
+      val totalCores = Try(totalExecutorCores.toInt).getOrElse(-1)
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: What is the purpose of the Try block here? if it fails the input is invalid, and proceeding with -1 can't be right.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 2]
```diff
@@ -658,19 +658,22 @@ private[deploy] class Master(
   private def startExecutorsOnWorkers(): Unit = {
     // Right now this is a very simple FIFO scheduler. We keep trying to fit in the first app
     // in the queue, then the second app, etc.
-    for (app <- waitingApps if app.coresLeft > 0) {
+    for (app <- waitingApps) {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Shouldn't all of this change be reverted?

**Review by srowen — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala (hunk 1 above)
Comment (eatoncys) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: Ok, I will remove the Try block, thanks.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala (hunk 2 above)
Comment (eatoncys) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Is not be better to compare app.coresLeft whih coresPerExecutor? If the coresLeft less than coresPerExecutor, it will return directly

Comment (eatoncys) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: @srowen If the total cores is not divisible by cores per executor, the compare app.coresLeft>0 will be always true, so it is better to compare app.coresLeft with coresPerExecutor than compare with 0.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala [hunk 3]
```diff
@@ -278,6 +278,12 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
     if (pyFiles != null && !isPython) {
       SparkSubmit.printErrorAndExit("--py-files given but primary resource is not a Python script")
     }
+    if (totalExecutorCores != null && executorCores != null
+        && (totalExecutorCores.toInt % executorCores.toInt) != 0) {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: Minor, but I think the parentheses are in the wrong place. They should go around the expression, or probably just be removed. You might avoid duplicating the mod expression for clarity.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala (hunk 2 above)
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: I see, Jerry was not saying that this part results in a logic change. This is OK.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 4]
```diff
@@ -658,19 +658,22 @@ private[deploy] class Master(
...
-          worker.coresFree >= coresPerExecutor.getOrElse(1))
-        .sortBy(_.coresFree).reverse
-      val assignedCores = scheduleExecutorsOnWorkers(app, usableWorkers, spreadOutApps)
-
-      // Now that we've decided how many cores to allocate on each worker, let's allocate them
-      for (pos <- 0 until usableWorkers.length if assignedCores(pos) > 0) {
-        allocateWorkerResourceToExecutors(
-          app, assignedCores(pos), coresPerExecutor, usableWorkers(pos))
+      // If the cores left is less than the coresPerExecutor,the cores left will not be allocated
+      if (app.coresLeft >= coresPerExecutor.getOrElse(1)) {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Might avoid duplicating `coresPerExecutor.getOrElse(1)` by referring to it as `requestedCoresPerExecutor` or something

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala (hunk 3 above)
Comment (eatoncys) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: Ok, I've modified the  mod expression repeated  to a val param and reused it,thanks.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala (hunk 4 above)
Comment (eatoncys) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Ok, I've modified the expression "val coresPerExecutor = app.desc.coresPerExecutor" to "val coresPerExecutor = app.desc.coresPerExecutor.getOrElse(1)" and reused it. And then using "app.desc.coresPerExecutor" directly in the function allocateWorkerResourceToExecutors. Thanks.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala [hunk 5]
```diff
@@ -278,6 +278,14 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
     if (pyFiles != null && !isPython) {
       SparkSubmit.printErrorAndExit("--py-files given but primary resource is not a Python script")
     }
+    if (totalExecutorCores != null && executorCores != null) {
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: Think about this again, I think this part of code logic could be moved to `SparkConf`, in case user sets these configuration via `SparkConf` object in a programmatic way.

**Comment by jerryshao** [comment]
@eatoncys can you please add a unit test in `MasterSuite` to verify your new code if possible?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala (hunk 5 above)
Comment (eatoncys) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: Ok, I have moved it to SparkConf, would you like to review it again, thanks.

**Comment by eatoncys** [comment]
@jerryshao, I have added a unit test in MasterSuite, would you like to review it again, thanks.

File: core/src/main/scala/org/apache/spark/SparkConf.scala [hunk 6]
```diff
@@ -543,6 +543,30 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging with Seria
       }
     }
 
+    if (contains("spark.cores.max")) {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: I think these checks for negative numbers are redundant with arg checking for `spark-submit`?

**Review by srowen — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/apache/spark/SparkConf.scala (hunk 6 above)
Comment (eatoncys) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: @srowen  Users may set these configuration via SparkConf object in a programmatic way, can I move the checkings from spark–submit to here, or removed it here directly,which is better?

File: core/src/main/scala/org/apache/spark/SparkConf.scala [hunk 7]
```diff
@@ -543,6 +543,30 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging with Seria
...
+      val totalCores = getInt("spark.cores.max", -1)
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: Please use `Option` instead, I think we don't need to check them if they're not set. Also looks like if we don't set this configuration, will it always lead to exception as default value is -1.

File: core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala [hunk 8]
```diff
@@ -704,6 +707,43 @@ class MasterSuite extends SparkFunSuite
   private def getState(master: Master): RecoveryState.Value = {
     master.invokePrivate(_state())
   }
+
+  test("Total cores is not divisible by cores per executor") {
```
Comment (jerryshao) [inline core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala]: IIUC I don't think this UT really reflects the code you changed in `Master`, even with the original code this two UTs should also be passed. I think you should test if the logics you changed is executed or not.

Comment (eatoncys) [inline core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala]: @jerryshao The result is same before and after my change, how to test them differently, any suggestion? thanks.

File: core/src/main/scala/org/apache/spark/SparkConf.scala (hunk 7 above)
Comment (eatoncys) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: @jerryshao I don't understand very cleanly, if we don't set this configuration ,the "if (contains("spark.cores.max")) " will not got into.

File: core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala (hunk 8 above)
Comment (jerryshao) [inline core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala]: My original thinking is that if it is possible to test wether the `if` branch is executed or not using `mock` and `verify`. But looks like it is not easy to test. Can you please investigate.

File: core/src/main/scala/org/apache/spark/SparkConf.scala (hunk 7 above)
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: Sorry my fault, I misunderstood the code.

File: core/src/main/scala/org/apache/spark/SparkConf.scala [hunk 9]
```diff
@@ -543,6 +543,30 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging with Seria
...
+      }
+    }
+    if (contains("spark.executor.cores")) {
+      val executorCores = getInt("spark.executor.cores", -1)
+      if (executorCores <= 0) {
+        throw new IllegalArgumentException(s"spark.executor.cores " +
+          s"(was ${get("spark.executor.cores")}) can only be a positive number")
+      }
+    }
+    if (contains("spark.cores.max") && contains("spark.executor.cores")) {
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: I think we could move the negative check to here to simplify the code.

File: core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala (hunk 8 above)
Comment (eatoncys) [inline core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala]: @jerryshao I have not any good way to test like this, any good suggestion? @srowen

File: core/src/main/scala/org/apache/spark/SparkConf.scala (hunk 9 above)
Comment (eatoncys) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: @jerryshao I put the negative check here first, but I think the app should  exit directly if the cores is  negative, so I move them out. And @srowen thinks these checks for negative numbers are redundant with arg checking for spark-submit, it may be a good way to move the checkings from spark-submit to here.

File: core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala (hunk 8 above)
Comment (jerryshao) [inline core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala]: I see, if there's no better way to verify it, I think it is not useful to add this two UTs.

File: core/src/main/scala/org/apache/spark/SparkConf.scala [hunk 10]
```diff
@@ -543,6 +545,42 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging with Seria
...
+        throw new IllegalArgumentException(s"spark.driver.memory " +
+          s"(was ${driverMemory}) can only be a positive number")
+      }
+    }
+    if (contains("spark.executor.memory")) {
+      val executorMemory = get("spark.executor.memory")
+      if (Try(JavaUtils.byteStringAsBytes(executorMemory)).getOrElse(-1L) <= 0) {
+        throw new IllegalArgumentException(s"spark.executor.memory " +
+          s"(was ${executorMemory}) can only be a positive number")
+      }
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: This above two checks seems unnecessary, let's not change unrelated code.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala [hunk 11]
```diff
@@ -258,23 +256,7 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
...
-    }
-    if (executorCores != null && Try(executorCores.toInt).getOrElse(-1) <= 0) {
-      SparkSubmit.printErrorAndExit("Executor cores must be a positive number")
-    }
-    if (totalExecutorCores != null && Try(totalExecutorCores.toInt).getOrElse(-1) <= 0) {
-      SparkSubmit.printErrorAndExit("Total executor cores must be a positive number")
-    }
-    if (numExecutors != null && Try(numExecutors.toInt).getOrElse(-1) <= 0) {
-      SparkSubmit.printErrorAndExit("Number of executors must be a positive number")
-    }
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: The above changes are valid and useful, I'd suggest to not change it.

File: core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala (hunk 8 above)
Comment (eatoncys) [inline core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala]: @jerryshao  Ok, I have removed them out, thanks

File: core/src/main/scala/org/apache/spark/SparkConf.scala (hunk 10 above)
Comment (eatoncys) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: @jerryshao  Ok ,I have moved them back, thanks

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala (hunk 11 above)
Comment (eatoncys) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: @jerryshao  Ok ,I have moved them back, thanks

**Comment by jerryshao** [comment]
@jiangxb1987 can you please help to review this PR? This is a simple code improvement to avoid some unnecessary code execution when left cores is not enough for one executor. 

I don't strong inclination on this PR since previous code also does correct behavior, I'd like to hear your thoughts.

**Comment by eatoncys** [comment]
cc @srowen

File: core/src/main/scala/org/apache/spark/SparkConf.scala [hunk 12]
```diff
@@ -543,6 +543,30 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging with Seria
...
+    }
+    if (contains("spark.executor.cores")) {
+      val executorCores = getInt("spark.executor.cores", -1)
+      if (executorCores <= 0) {
+        throw new IllegalArgumentException(s"spark.executor.cores " +
+          s"(was ${get("spark.executor.cores")}) can only be a positive number")
+      }
+    }
+    if (contains("spark.cores.max") && contains("spark.executor.cores")) {
+      val totalCores = getInt("spark.cores.max", 1)
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: @jerryshao I think most of the new argument checking code is redundant. I think the 7 lines from here are all that are needed

Comment (eatoncys) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: @srowen Users may set these configuration via SparkConf object in a programmatic way after arg checking for spark-submit. For example, an app code may be like this: "val conf = new SparkConf().set("spark.executor.cores","-1")" before SparkContext created, the checkings here can get the error and exit directly.

File: core/src/main/scala/org/apache/spark/SparkConf.scala [hunk 13]
```diff
@@ -543,6 +543,30 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging with Seria
       }
     }
 
+    if (contains("spark.cores.max")) {
+      val totalCores = getInt("spark.cores.max", -1)
+      if (totalCores <= 0) {
+        throw new IllegalArgumentException(s"spark.cores.max (was ${get("spark.cores.max")})" +
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: ${get("spark.cores.max")} => $totalCores ?

File: core/src/main/scala/org/apache/spark/SparkConf.scala [hunk 14]
```diff
@@ -543,6 +543,30 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging with Seria
...
+          s" can only be a positive number")
+      }
+    }
+    if (contains("spark.executor.cores")) {
+      val executorCores = getInt("spark.executor.cores", -1)
+      if (executorCores <= 0) {
+        throw new IllegalArgumentException(s"spark.executor.cores " +
+          s"(was ${get("spark.executor.cores")}) can only be a positive number")
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: ditto

File: core/src/main/scala/org/apache/spark/SparkConf.scala (hunk 12 above)
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: Let's just keep this checking branch, or move the previous negative checking code here.

**Review by jiangxb1987 — COMMENTED** [review]
The changes looks good overall, only a few comments.

File: core/src/main/scala/org/apache/spark/SparkConf.scala (hunk 12 above)
Comment (eatoncys) [inline core/src/main/scala/org/apache/spark/SparkConf.scala]: @srowen @jiangxb1987 Ok, I have removed the argument checking code, thanks.

**Comment by jerryshao** [comment]
LGTM!

**Review by srowen — APPROVED** [review]

**Review by jiangxb1987 — APPROVED** [review]
LGTM, cc @cloud-fan

**Comment by cloud-fan** [comment]
LGTM, merging to master!
